File: 2328-number-of-increasing-paths-in-a-grid/2328-number-of-increasing-paths-in-a-grid.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        r,c=len(grid),len(grid[0])
        dp=[[0 for _ in range(c)] for _ in range(r)]
        dx=[-1,0,1,0]
        dy=[0,-1,0,1]
        mod=10**9+7
        
        def dfs(x,y):
            if dp[y][x]:
                return dp[y][x]
            ret = 1
            for i in range(4):
                nx=x+dx[i]
                ny=y+dy[i]
                if not (0<=nx<c and 0<=ny<r):
                    continue
                if grid[ny][nx] > grid[y][x]:
                    ret += dfs(nx,ny) 
            dp[y][x]=ret%mod
            return ret
        for y in range(r):
            for x in range(c):
                dfs(x,y)
        answer=0
        for y in range(r):
            for x in range(c):
                answer+=dp[y][x]
        return answer%mod
                    
            
            
        

File: 2328-number-of-increasing-paths-in-a-grid/2328-number-of-increasing-paths-in-a-grid.rs
impl Solution {
    pub fn dfs(grid:&Vec<Vec<i32>>, dp:&mut Vec<Vec<i32>>, x:usize,y:usize,dx:&[isize;4],dy:&[isize;4])->i32{
        if dp[y][x] != 0 {
            return dp[y][x];
        }
        let mut ret = 1;
        for i in 0..4{
            let nx = ((x as isize) + dx[i]) as usize;
            let ny = ((y as isize) + dy[i]) as usize;
            if !(0 <= nx && nx < grid[0].len() && 0 <= ny && ny < grid.len()){ continue; }
            if grid[y][x] >= grid[ny][nx] {
                continue;
            }
            ret += Solution::dfs(grid, dp, nx, ny, dx, dy) % 1_000_000_007;
        }
        dp[y][x]=ret % 1_000_000_007;
        return ret;
    }
    pub fn count_paths(grid: Vec<Vec<i32>>) -> i32 {
        let (r,c) = (grid.len(),grid[0].len());
        let mut dp = vec![vec![0;c];r];
        let dx:[isize;4] = [-1,0,1,0];
        let dy:[isize;4] = [0,-1,0,1];
        
        for y in 0..r{
            for x in 0..c{
                Solution::dfs(&grid,&mut dp, x, y, &dx, &dy);
            }
        }
        let mut answer = 0;
        
        for y in 0..r{
            for x in 0..c{
                answer += dp[y][x];
                answer %= 1_000_000_007;
            }
        }
        answer % 1_000_000_007
    }
}